(AB#145508) Update config conceptual doc (#179)
MicrosoftDocs/PowerShell-Docs-DSC: dsc/docs-conceptual/dsc-3.0/concepts/configurations.md

ms.date: 08/07/2023
title: DSC Configuration Documents
---

# DSC Configuration Documents

<!-- markdownlint-disable MD013 -->

In Microsoft's Desired State Configuration (DSC) platform, DSC configuration documents declare the
desired state of a system as data files. Configuration documents define a collection of
[DSC Resource][01] instances to describe what the desired state should be, not how to put the

> system into that state. The DSC Resources handle the "how" for the configuration document's

Reviewer: ```suggestion
system into that state. The DSC Resources handle the _how_ for the configuration document
```

- `resources` (required) - Defines the collection of resource instances the document manages.
- `metadata` (optional) - Defines an arbitrary set of annotations for the document. DSC doesn't
  validate this data or use it directly. The annotations can include notes like who authored the
  document, the last time someone updated it, or any other information. DSC doesn't use the
  annotations. The metadata is for documentation or other tools to use.
- `parameters` (optional) - Defines a set of runtime options for the configuration. Parameters can
  be referenced by resource instances to reduce duplicate definitions or enable dynamic values.
  Parameters can have default values and can be set on any configuration operation.
- `variables` (optional) - Defines a set of reusable values for the configuration. Variables can be
  referenced by resource instances to reduce duplicate definitions.

> [!NOTE]
> DSC isn't implemented to process the `parameters` and `variables` properties yet. They can be
> defined in a document, but not referenced.

## Defining a configuration document

Minimally, a configuration document defines the `$schema` and `resources` properties. The

> `$schema` property must be a valid URI for DSC's configuration document schema. The `resources`

Reviewer: ```suggestion
`$schema` property must be a valid URI for the DSC configuration document schema. The `resources`
```

property must define at least one DSC Resource instance.

For example:

```yaml
# example.dsc.config.yaml
$schema: https://raw.githubusercontent.com/PowerShell/DSC/main/schemas/2023/08/config/document.json
resources:
  - name: example key value
    type: Microsoft.Windows/Registry
    properties:
      keyPath:   HKCU\example\key
      valueName: Example
      valueData:
        String: This is an example.
```

> The example document declares a single resource instance called `example key value`. The instance
> uses the `Microsoft.Windows/Registry` resource to declare that:

Reviewer: ```suggestion
The example document defines a single resource instance called `example key value`. The instance
uses the `Microsoft.Windows/Registry` resource to declare the following desired state:
```

- The `example\key` registry key should exist in the system's current user hive.
- The `example\key` registry key should have a value called `Example`.
- The `Example` value should be the string `This is an example`.

The example document is _declarative_. It describes the desired state, not how to put the system
into that state. It relies on the `Microsoft.Windows/Registry` resource to handle getting, testing,
and setting the state of the registry key and value.

### Defining resource instances

As shown in the prior example, configuration documents include a collection of resource instances.

> Together, the instances describe a system's desired state. A configuration document can include any
> number of resource instances. The instances can be of any available DSC Resource.

Reviewer: ```suggestion
Together, the instances describe the desired state of a system. A configuration document can include any
number of resource instances.
```

I am not sure what that last sentence (removed) was supposed to mean.

Author: Trying to clarify that you're not limited to defining instances of a single resource type at a time - agree with cutting though.

A resource instance declaration always includes:

- `name` - A short, human-readable name for the instance that's unique in the document. This name

>   is used for logging and it helps to document an instance's purpose in the document.
> - `type` - The [fully qualified type name][02] for the resource to identify the resource DSC should
>   use to manage the instance.

Reviewer: ```suggestion
  is used for logging and it helps describe the purpose of the instance.
- `type` - The [fully qualified type name][02] of the resource that DSC should use to manage the instance.
```

Are these 2 uses of "instance" talking about the same thing?

Author: Yes - the _instance_ is the arbitrary component being managed. It's defined in three parts:

- The `properties` keyword defines the desired state of the instance.
- The `type` keyword defines the resource that processes the instance. The processing includes both property validation and operation invocation.
- The `name` keyword defines an annotation for human readers and users of the configuration. The name is included in logging for the instance and serves as minimal documentation for the instance.

- `properties` - The desired state for the instance. DSC validates the values against the
  resource's instance schema.

Configuration documents can't include the same instance more than once. Declaring the same instance
with different properties leads to enforcement cycling, where each declaration enforces an
incompatible state for the instance on every run.

## Getting the current state of a configuration

The `dsc config get` command retrieves the current state of the resource instances defined in a
configuration document. DSC also collects any message emitted by the resources during the operation
and indicates whether any of the resources raised an error.

```sh
cat ./example.config.dsc.yaml | dsc config get
```

```yaml
results:
- name: example key value
  type: Microsoft.Windows/Registry
  result:
    actualState:
      $id: https://developer.microsoft.com/json-schemas/windows/registry/20230303/Microsoft.Windows.Registry.schema.json
      keyPath: ''
messages: []
hadErrors: false
```

## Testing a configuration

> The `dsc config test` command compares the current state of the resource instances defined in a
> configuration document to their desired state. The result for each instance includes:

Reviewer: ```suggestion
The `dsc config test` command compares the current state of the resource instances to their desired state. The result for each instance includes:
```
